enhancement: support revert receipts

rvmelkonian opened the pull request:
Closes #669 

## Changelog
• updates codegen macro, schema, handler, web-api and creates new test for indexing Revert reciepts. 
• adds `pub enum FuelError` to `packages/fuel-indexer-types/src/abi.rs`

## Note
- When trying to store the `ra` value in the db, we get a postgres out of bounds error. The ra value emitted in the receipt `18446744073709486084`, is too large to store in the db. So I convert that to the equivalent value from the VM `PanicReason::Revert` found [here](https://github.com/FuelLabs/fuel-vm/blob/master/fuel-asm/src/panic_reason.rs). These seemed like a good solution, as the user could still see what error specifically was indexed, in this case a revert. 

- If we don't want developers to manually implement this error conversion in their indexer, we could create a module that they could use. I'll put up an issue.

## Testing 
Run the new test `cargo test test_can_trigger_and_index_panic_function_postgres --features e2e,postgres`

deekerno reviewed (changes requested):
@rvmelkonian You can represent the `u64` type with a `UInt8`. 👌🏽 You can also check [this section](https://fuellabs.github.io/fuel-indexer/master/reference-guide/data-types/types.html) of the indexer book for equivalent data types between Sway and GraphQL. Furthermore, you can check the receipt spec for descriptions for the other receipt types and see how entities represent receipt data fields of type `u64`.

Other than the missing `val` field, it looks good! Great job!

rvmelkonian commented:
@deekerno Nice. That table in our docs is very useful.

deekerno commented on packages/fuel-indexer-types/src/abi.rs:
```diff
@@ -155,3 +189,17 @@ impl NativeFuelType for Call {
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct Revert {
+    pub contract_id: ContractId,
+    pub ra: u64,
```
What if we changed this to something like "error_val" in order to give users a more descriptive name? It's technically `ra` for "register A" but it would be nice to communicate the intended usage.

deekerno commented on packages/fuel-indexer-types/Cargo.toml:
```diff
@@ -11,6 +11,7 @@ description = "Fuel Indexer Types"
 
 [dependencies]
 chrono = { version = "0.4", features = ["serde"] }
+fuel-abi-types = "0.2.1"
```
Since you're not using this crate anymore, you can remove this.

deekerno commented on packages/fuel-indexer-macros/src/indexer.rs:
```diff
@@ -509,6 +509,12 @@ fn process_fn_items(
                                 let data = bincode::serialize(&abi::TransferOut{ contract_id: id, to, asset_id, amount, pc, is }).expect("Bad encoding,");
                                 decoder.decode_type(ty_id, data);
                             }
+                            Receipt::Revert { id, ra, .. } => {
+                                #contract
+                                let ty_id = abi::Revert::type_id();
+                                let data = bincode::serialize(&abi::Revert{ contract_id: id, ra }).expect("Bad encoding,");
+                                decoder.decode_type(ty_id, data);
+                            }
```
Just a thought, but what if we moved the bitmasking from the handler to the codegen itself? For example (using the other suggestion of an `error_val` field in `abi::Revert`):
```
Receipt::Revert { id, ra, .. } => {
    #contract
    let ty_id = abi::Revert::type_id();
    let data = bincode::serialize(&abi::Revert{ contract_id: id, error_val: u64::from(ra & 0xF) }).expect("Bad encoding,");
    decoder.decode_type(ty_id, data);
}
```

It seems improbable that a user would use the very large decimal values directly from the receipt other than to apply a bitmask upon them so if we we do it for them here in the codegen, they wouldn't have to worry about it. We could then add a mapping of the much smaller values and their meanings to a `Revert` page of the Receipts section of the book. Additionally, they could also store the value straight into the database as well.

Furthermore, you wouldn't have to define a `FuelError` enum or implement the `unmask()` operation.

rvmelkonian replied:
Yes, you are right we shouldn't expose this to the user 👌.
I'll update the receipt section of the docs in this PR also.

deekerno commented on packages/fuel-indexer-tests/src/fixtures.rs:
```diff
@@ -644,6 +644,28 @@ pub mod test_web {
+
+        println!("Result: {:?}", result);
+
+        match result {
+            Ok(_) => HttpResponse::Ok(),
+            Err(_) => {
+                //Force 200 on err response here, because we have successfully triggered a panic
+                HttpResponse::Ok()
+            }
+        }
```
Nice catch here! I think you can get away with removing the comment and result printing, and then matching on just the exhaustive case since we're doing the same action either way. For example:
```
    enum MysteryIncSounds {
        Zoinks,
        Jinkies,
        RuhRoh
    }
    
    let mystery_inc_sound = MysteryIncSounds::Zoinks;
    
    match mystery_inc_sound {
        _ => println!("Why are kids investigating crimes anyways?"),
    }
```

rvmelkonian replied:
This is a vivd example, love it.

rvmelkonian replied:
turns out this is fine 
```rust
async fn fuel_indexer_test_trigger_revert(
        state: web::Data<Arc<AppState>>,
    ) -> impl Responder {
        let _ = state
            .contract
            .methods()
            .trigger_revert()
            .tx_params(tx_params())
            .call()
            .await;

        HttpResponse::Ok()
    }
    ```
    I initially added the match as I was getting error response codes, but it must have been for some other reason.

deekerno commented on packages/fuel-indexer-types/src/abi.rs:
```diff
@@ -6,6 +6,40 @@ pub use fuel_tx::Receipt;
 pub use fuels_types::Identity;
 use serde::{Deserialize, Serialize};
 
+#[repr(u64)]
+pub enum FuelError {
+    FailedRequire = 0xffff_ffff_ffff_0000_u64,
+    FailedTransferToAddress = 0xffff_ffff_ffff_0001_u64,
+    FailedSendMessage = 0xffff_ffff_ffff_0002_u64,
+    FailedAssertEq = 0xffff_ffff_ffff_0003_u64,
+    FailedAssert = 0xffff_ffff_ffff_0004_u64,
+}
+
+impl From<u64> for FuelError {
+    fn from(val: u64) -> Self {
+        match val {
+            0xffff_ffff_ffff_0000_u64 => FuelError::FailedRequire,
+            0xffff_ffff_ffff_0001_u64 => FuelError::FailedTransferToAddress,
+            0xffff_ffff_ffff_0002_u64 => FuelError::FailedSendMessage,
+            0xffff_ffff_ffff_0003_u64 => FuelError::FailedAssertEq,
+            0xffff_ffff_ffff_0004_u64 => FuelError::FailedAssert,
+            _ => panic!("Invalid FuelError value: {}", val),
+        }
+    }
+}
+
+impl FuelError {
+    pub fn unmask(&self) -> u64 {
+        match self {
+            FuelError::FailedRequire => 0x0000,
+            FuelError::FailedTransferToAddress => 0x0001,
+            FuelError::FailedSendMessage => 0x00002,
+            FuelError::FailedAssertEq => 0x0003,
+            FuelError::FailedAssert => 0x0004,
+        }
+    }
+}
+
```
See other comment about doing bitmasking in the codegen phase.

deekerno reviewed (changes requested):
Looking good! Just a few thoughts.

rvmelkonian commented:
@deekerno I've addressed you're feedback, yeah it's much better just doing `error_val: u64::from(ra & 0xF)` in the codegen! Also, be sure to run `mdbook serve` as I have updated the book. I also re-ordered the items of `7.2 Receipts` so that they are in alphabetical order.

deekerno commented on docs/src/reference-guide/indexing/receipts/revert.md:
````diff
@@ -0,0 +1,31 @@
+
+```rust, ignore
+fn handle_revert(revert: abi::Revert) {
+  // handle the transfer out
````
This should be `// handle the revert`.

rvmelkonian replied:
Whoops, thanks.

deekerno reviewed (changes requested):
Left some final thoughts. Once they're addressed, I'll approve. ✅

rvmelkonian commented on packages/fuel-indexer-types/src/abi.rs:
```diff
@@ -155,3 +155,17 @@ impl NativeFuelType for Call {
+    pub error_val: u64,
+}
+
+pub struct SomeThing();
```
My apologies for leaving it! I was experimenting and left it in.

deekerno reviewed (approved):
Thanks for adding some documentation! Nice job on this!

🚢
